### DISCUSSION TITLE
Dedicated point cloud primitive?

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#621] [Ideas]
I've been testing with a big set (1000+) of large point clouds (400k points). In essence it's an animated point cloud, drawing only one at any time. The existing point cloud demo, using the "uv-based quad" was a big help! However, memory usage is quite high, due to the amplification of having to use 4 vertices + 6 indices per point. One way to reduce memory is to not store a per-vertex normal as it's not used in the shader. Plus, instead of always storing the same u/v +/-0.5 values over four vertices we could use a 2-bit value in each vertex (stored as either a byte or uint16) and use that in the vertex shader to compute the uv offsets. 

The above would lead to 18 bytes per vertex (12b pos, 2b uv, 4b col) instead of the current 36 bytes per vertex (12b pos, 12b nor, 8b uv, 4b col), ie 50% less memory. However, this would need a dedicated point cloud primitive as I'm sure you would want to keep `Mesh` and `vert_t` unchanged. 

@maluoi Before I even begin to look into *if* I want to start implementing a new point cloud primitive (at both C and C++ levels) as described above, do you see the value and do you agree it's the best approach? I have no idea if there's better ways to do point rendering, i.e. without the uv-quad trick, that is portable to the platforms you target SK for.

**Comment by maluoi** [comment]
The expanded quad is a nice trick that brings some decent efficiency, but you're right in that it's definitely not the most optimal implementation! There's a couple additional tricks that can be done to make this even better, and you've listed the big ones. (UVs may be infer-able from vertex index) There should be a collection of ways to do a lot of this with compute shaders too.

The big reason why the memory things weren't done in the first place is that StereoKit's renderer doesn't have support for different vertex formats! There's a surprising amount of architectural work necessary to coordinate different formats, and I haven't found the time to fully add this yet. I was working on this [not so long ago in a branch](https://github.com/StereoKit/sk_gpu/tree/feature/custom_formats), I got as far as custom index formats and was working through vertex stuff, but I don't recall exactly where I stopped. There were changes needed all the way as far down as the shader compiler!

It probably doesn't make sense to make any CPU side optimizations without also having the custom vertex formats backing that on the GPU, so that'll need to get done first. But when that does get finished, this should be easily doable using the exiting mesh assets no problem!
